File: typescript/sort/merge.ts
//import * as check from "./check.ts";

export const mergeSort = new class MergeSort {
    operations = 0;

    #nMerge(left: number[], right: number[]): number[] {
        let j = 0, i = 0;
        const s: number[] = [];
        while (s.length != (left.length + right.length)) {
            this.operations++;
            if (left[j] !== undefined && right[i] !== undefined) s.push(left[j] <= right[i] ? left[j++] : right[i++]);
            else if (left[j] !== undefined) s.push(left[j++]);
            else if (right[i] !== undefined) s.push(right[i++]);
        };
        //console.log("merged,left,right",s,left,right);
        return s;
    };
    #nSplit(arr: number[]): number[] {
        if (arr.length <= 1) return [...arr];
        const middle = Math.floor(arr.length / 2);
        const left = this.#nSplit(arr.slice(0, middle));
        const right = this.#nSplit(arr.slice(middle));
        this.operations++;
        return this.#nMerge(left, right);
    };
    number(arr: number[]): number[] {
        this.operations = 0;
        return this.#nSplit(arr);
    };
};

export const mergeSortDebug = new class MergeSortDebug {
    workArr: number[] = [];

    // Hooks for visualization
    async next() { return new Promise(r => setTimeout(r)); }
    async checked() { return new Promise(r => setTimeout(r)); }
    async get(index: number) { return this.workArr[index]; }
    async set(index: number, value: number) { return this.workArr[index] = value; }

    async #nMerge(start: number, mid: number, end: number): Promise<void> {
        const left = this.workArr.slice(start, mid);
        const right = this.workArr.slice(mid, end);
        let i = 0, j = 0, k = start;

        while (i < left.length && j < right.length) {
            await this.checked();
            if (left[i] <= right[j]) {
                await this.set(k++, left[i++]);
            } else {
                await this.set(k++, right[j++]);
            }
            await this.next();
        }

        while (i < left.length) {
            await this.set(k++, left[i++]);
            await this.next();
        }

        while (j < right.length) {
            await this.set(k++, right[j++]);
            await this.next();
        }
    }

    async #nSplit(start: number, end: number): Promise<void> {
        if (end - start <= 1) return;

        const mid = Math.floor((start + end) / 2);
        await this.#nSplit(start, mid);
        await this.#nSplit(mid, end);
        await this.#nMerge(start, mid, end);
    }

    async number(arr: number[]): Promise<number[]> {
        this.workArr = [...arr];
        await this.#nSplit(0, this.workArr.length);
        return this.workArr;
    }
}

File: rust_sort/src/sort/merge.rs
fn number_merge<T>(left: Vec<T>, right: Vec<T>)->Vec<T>
where T: PartialEq + PartialOrd + Copy,
{
    let mut j=0;
    let mut i=0;
    let ll=left.len();
    let rl=right.len();
    let nl=ll+rl;
    let mut s:Vec<T>=Vec::new();
    while s.len()!=nl {
        if ll>j&&rl>i {s.push(if left[j]<=right[i]{j+=1;left[j-1]}else{i+=1;right[i-1]});}
        else if ll>j {j+=1;s.push(left[j-1]);}
        else if rl>i {i+=1;s.push(right[i-1]);}
    };
    //console.log("merged,left,right",s,left,right);
    return s;
}

pub fn number<T>(arr: Vec<T>)->Vec<T>
where T: PartialEq + PartialOrd + Copy,
{
    if arr.len()<=1 {return arr;}
    let middle=arr.len()/2;
    let left=number(arr[0..middle].to_vec());
    let right=number(arr[middle..arr.len()].to_vec());
    return number_merge(left,right);
}